crates/matrix-sdk-appservice/src/lib.rs

```diff
@@ -118,6 +118,116 @@ pub type Port = u16;
 
 const USER_KEY: &[u8] = b"appservice.users.";
 
+/// Builder for a virtual user
+#[derive(Debug)]
+pub struct VirtualUserBuilder<'a> {
+    appservice: &'a AppService,
+    localpart: &'a str,
+    device_id: Option<OwnedDeviceId>,
+    client_builder: ClientBuilder,
+    log_in: bool,
+    restored_session: Option<Session>,
+}
+
+impl<'a> VirtualUserBuilder<'a> {
+    /// Create a new virtual user builder
+    /// # Arguments
+    ///
+    /// * `localpart` - The localpart of the virtual user
+    pub fn new(appservice: &'a AppService, localpart: &'a str) -> Self {
+        Self {
+            appservice,
+            localpart,
+            device_id: None,
+            client_builder: Client::builder(),
+            log_in: false,
+            restored_session: None,
+        }
+    }
+
+    /// Set the device id of the virtual user
+    pub fn with_device_id(mut self, device_id: Option<OwnedDeviceId>) -> Self {
```

Comment:
Builder looks good!

I'd remove the `with_` from all builder methods - that's unnecessary duplication.